quixstreams/kafka/configuration.py

```diff
@@ -0,0 +1,147 @@
+
+from pydantic import AliasGenerator, SecretStr, AliasChoices, Field
+from pydantic.functional_validators import BeforeValidator
+from pydantic_settings import BaseSettings, SettingsConfigDict
+from typing_extensions import Self, Annotated
+
+__all__ = ("ConnectionConfig",)
+
+MECHANISM_ALIAS = AliasChoices("sasl_mechanism", "sasl_mechanisms")
+
+
+class ConnectionConfig(BaseSettings):
+    """
+    Provides an interface for all librdkafka connection-based configs.
+
+    Allows converting to or from a librdkafka dictionary.
+
+    Also obscures secrets and handles any case sensitivity issues.
+    """
+
+    model_config = SettingsConfigDict(
+        alias_generator=AliasGenerator(
+            # used during model_dumps
+            serialization_alias=lambda field_name: field_name.replace("_", "."),
+        ),
+    )
+
+    @classmethod
+    def secret_fields(cls) -> List[str]:
+        """
+        Get all the fields that are of the type "SecretStr" (passwords)
+
+        :return: a list of secret field names
+        """
+        fields = []
+        for name, info in ConnectionConfig.model_fields.items():
+            if SecretStr in get_args(info.annotation):
+                fields.append(name)
+        return fields
+
+    @classmethod
+    def from_librdkafka_dict(cls, config: dict, ignore_extras: bool = False) -> Self:
+        """
+        Create a `ConnectionConfig` from a librdkafka config dictionary.
+
+        :param config: a dict of configs (like {"bootstrap.servers": "url"})
+        :param ignore_extras: Ignore non-connection settings (else raise exception)
+
+        :return: a ConnectionConfig
+        """
+        config = {name.replace(".", "_"): v for name, v in config.items()}
+        if ignore_extras:
+            valid_keys = set(cls.model_fields.keys()) | set(MECHANISM_ALIAS.choices)
+            config = {k: v for k, v in config.items() if k in valid_keys}
+        return cls(**config)
+
+    def as_librdkafka_dict(self) -> dict:
+        """
+        Dump any non-empty config values as a librdkafka dictionary.
+
+        >***NOTE***: Dumps all secret fields in plaintext.
+
+        :return: a librdkafka-compatible dictionary
+        """
+        dump = self.model_dump(by_alias=True, exclude_none=True)
+        for field in self.secret_fields():
+            field = field.replace("_", ".")
+            if dump.get(field):
+                dump[field] = dump[field].get_secret_value()
+        return dump
+
+    def as_printable_json(self, indent: int = 2) -> str:
+        """
+        Dump any non-empty config values as a librdkafka-formatted JSON string.
+
+        Safely obscures the any "secret" fields (passwords).
+
+        :param indent: the indent of the JSON
+
+        :return: a string-dumped JSON
+        """
+        return self.model_dump_json(by_alias=True, exclude_none=True, indent=indent)
+
+    def __str__(self) -> str:
+        return self.as_printable_json()
```

Comment:
I'm unsure if returning JSON in `__str__` is appropriate here because JSON contains additional formatting and linebreaks.

I suggest using a dict instead and cast it to a string:
```
>>> str(self.model_dump(by_alias=True, exclude_none=True))
# Prints "{'bootstrap.servers': '123', 'sasl.password': SecretStr('**********')}"
```